Establish survey response structure

asheren opened the pull request:
+ Establishes route to get to survey response
+ Submit a new survey response
+ Write related feature specs
+ Establish survey response model, and related specs
+ Establish survey response controller with necessary CRUD methods

TO DO:
+ create helper method (or enable capybara setup to allow for?) `sign_in` for feature specs
+ When trying to run `save_and_open_screenshot`, I got the error ` Capybara::NotSupportedByDriverError:
       Capybara::Driver::Base#save_screenshot`
Which makes me believe that I need to set up a driver or capybara settings a little differently that I currently have them set in the spec_helper file.

asheren committed:
added annotate and testing gems. removed spring.

asheren committed:
Annotates the user model

asheren committed:
Adds migration for survey responses.

Survey responses have a few fields.
1. perceived confidence level: this is the user's rating of how confident
they feel in their coding abilities at the time of taking the survey
2. material difficulty rating: this is the user's assessment of how difficult
the material is that they are working on
3. notes: notes are optional but helpful in adding context to why someone
may be feeling a certain way about either the material or their confidence level
4. timestamps: in particular because we want to be able to look at patterns
over different periods of time, the created_at stamp here is particularly
important

asheren committed:
TDD-ing this feature. This commit includes test set up.

asheren committed:
Creates and annotates survey responses table and model

asheren committed:
added validations for 2 survey questions and have passing specs.

asheren committed:
deletes comments

asheren committed:
formats postgres database

asheren committed:
Adds survey responses updated migration, validations, and tests

asheren committed:
fixed gemfile.lock conflicts

asheren committed:
Enables route and ability to fill out a new survey.

asheren committed:
New feature specs!!!

These aren't quite passing yet because of additional sign_in and capybara
set up that needs to happen but the structure is there.

asheren committed:
working on improving helpers and test setup. Test passes right now but I do not think it should.

asheren committed:
Adds authentication method

To ensure that a user must be logged in in order to fill out a survey, an
authentication method was defined.

asheren committed:
Creates authentication helper and sign in method

This allows the user to be signed in for the tests correctly

asheren committed:
Revises authenticate method

auth did not persist and so when the before_action was being called on the
survey responses controller, auth was nil even though there was a session[user.id].
To fix, we actually just need to check if the current user exists since the
current user method checks the session[:user_id]

csexton commented:
I would suggest rolling your own `sign_in`/`current_user` helpers fo the tests. There are some hoops to jump through, but think it would be good to go through the exercise. Gotta do a little meta-programming to pull it off nicely.

- Add test helpers for `sign_in` and `current_user`
- Insert the current user into yoru controller (perhaps by monkey patching ActionController)
- Reset after each spec is run

The helpers I use at work are open source, you can [see them here](https://github.com/RadiusNetworks/omniauth-radius-rails/blob/master/lib/kracken/rspec.rb). Also may want to check the the omniauth [wiki page on integration testing](https://github.com/omniauth/omniauth/wiki/Integration-Testing), however I found I didn't need to use this for my testing.

asheren committed:
Added create method

asheren committed:
Tests that survey has necessary information once the survey response is created

Also creates the index page since the create action redirects to index

asheren committed:
failing tests, why?

asheren commented:
@kopchickm specs are failing on the last commit. WOrks in the UI but not in the spec. Save and open page shows everything is fine.

```
Failures:

  1) Submitting a new code confidence survey A user can enter in a number for perceived code confidence
     Failure/Error: select("5", :from => "perceived_code_confidence")
     
     Capybara::ElementNotFound:
       Unable to find visible select box "perceived_code_confidence" that is not disabled
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:314:in `block in synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/base.rb:85:in `synchronize'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:302:in `synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:37:in `find'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/actions.rb:186:in `select'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/session.rb:808:in `block (2 levels) in <class:Session>'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/dsl.rb:50:in `block (2 levels) in <module:DSL>'
     # ./spec/features/survey_response_new_survey_spec.rb:11:in `block (2 levels) in <top (required)>'

  2) Submitting a new code confidence survey A user can enter in a number for material difficulty rating
     Failure/Error: select "2", :from => "material_difficulty_rating"
     
     Capybara::ElementNotFound:
       Unable to find visible select box "material_difficulty_rating" that is not disabled
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:314:in `block in synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/base.rb:85:in `synchronize'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:302:in `synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:37:in `find'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/actions.rb:186:in `select'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/session.rb:808:in `block (2 levels) in <class:Session>'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/dsl.rb:50:in `block (2 levels) in <module:DSL>'
     # ./spec/features/survey_response_new_survey_spec.rb:22:in `block (2 levels) in <top (required)>'

  3) Submitting a new code confidence survey A user can add notes if they would like
     Failure/Error: expect(survey_response.notes).to eq("Today my teammates were supportive and positive.")
     
     NoMethodError:
       undefined method `notes' for nil:NilClass
     # ./spec/features/survey_response_new_survey_spec.rb:41:in `block (2 levels) in <top (required)>'

  4) Submitting a new code confidence survey A user can submit the survey
     Failure/Error: select "5", :from => "perceived_code_confidence"
     
     Capybara::ElementNotFound:
       Unable to find visible select box "perceived_code_confidence" that is not disabled
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:314:in `block in synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/base.rb:85:in `synchronize'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:302:in `synced_resolve'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/finders.rb:37:in `find'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/node/actions.rb:186:in `select'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/session.rb:808:in `block (2 levels) in <class:Session>'
     # /Users/allisonmcmillan/.rvm/gems/ruby-2.3.1/gems/capybara-2.17.0/lib/capybara/dsl.rb:50:in `block (2 levels) in <module:DSL>'
     # ./spec/features/survey_response_new_survey_spec.rb:48:in `block (2 levels) in <top (required)>'

Finished in 8.57 seconds (files took 2.01 seconds to load)
4 examples, 4 failures
```
failures ^^
any thoughts on what's going on?

asheren committed:
booyah. passing specs. specs were not passing before because they were not saving correctly because of model validations.

asheren committed:
adds index view and controller action

asheren committed:
deletes commented code

asheren committed:
Adds index spec which affects other parts of the codebase

Adding the index spec shows a place where the code can be DRYer. This creates
a factory. However when the factory is created, since a survey response is
generated, the model spec needs to be updated to show invalid survey responses.

The index spec also needs to show results only for 1 user, so a user is
added to the factory.

asheren committed:
With the addition of user to the survey response factory, associations need updating as well

A survey response belongs to a user and a user has many survey responses.
This commit makes those associations clear.

asheren commented:
@kopchickm or @csexton I added a user to the survey response factory since writing that index spec showed me that a survey response belongs to a user and that association needs to be made clear.  in the last commit, I make that association clear, but I'm pretty sure I need to update the schema in order to reflect that, adding a `user_id` to they survey_responses table. With rails, I don't need to create a separate join table right? I can just create a migration that adds `user_id` to the `survey_responses` table?

asheren commented:
@kopchickm or @csexton I added a user to the survey response factory since writing that index spec showed me that a survey response belongs to a user and that association needs to be made clear.  in the last commit, I make that association clear, but I'm pretty sure I need to update the schema in order to reflect that, adding a `user_id` to they survey_responses table. With rails, I don't need to create a separate join table right? I can just create a migration that adds `user_id` to the `survey_responses` table?

asheren commented:
Did this and added an index so disregard the question

asheren commented:
Did this and added an index so disregard the question

csexton commented:
Not sure I 100% follow your question, but looking at [2eba5d7](https://github.com/asheren/code_mapping/pull/1/commits/2eba5d7092c194b2dd85c78b96bf6671de4084f9) it looks like you added the `belongs_to` and ja just need to add the `user_id` column to the database.

There is not foreign key constraints in Rails, it does it by convention. So as long as you have the right column name (in this case `user_id`) and the corresponding `belongs_to`/`has_many` in the models the relationship should work.

I'd test it in the Rails console to make sure everything is wired up properly.

asheren committed:
updates the schema

asheren committed:
connects survey response to the current user to complete the association

asheren committed:
Creates descending index list and index spec

Also deletes commented out code.
TODO spec still failing

asheren committed:
adds signout link and link to fill out a new survey.

asheren committed:
Adds ability for a user to delete a survey

Controller method and view layer

TODO spec is still failing

asheren commented on spec/features/survey_response_new_survey_spec.rb:
```diff
@@ -62,10 +62,13 @@
+    expect(page.first("tr:nth-of-type(2)")).to have_content(survey_response.perceived_code_confidence)
+    expect(page.first("tr:nth-of-type(2)")).to have_content(survey_response.material_difficulty_rating)
+    expect(page.first("tr:nth-of-type(2)")).to have_content(survey_response.notes)
+    expect(page.last("tr")).to have_content(survey_response.notes)
```
@kopchickm Running this spec doesn't pass because it's pulling in info from the last spec. So if I do a pry and check what `survey_response` and `survey_response_2` are, it is what's in the factory, but then the save and open page shows me that what's there is what has been generated from previous specs in the file. Additionally, if I run the spec in isolation, then the save_and_open_screenshot just shows me a blank index view (headers but no survey entries). Any thoughts or ideas?

kopchickm replied:
My first thought is that `User.last` isn't necessarily going to be the same thing every time, if there are multiple users in the system. You could start by making sure you're signed in as the same user that owns the 2 survey responses.

I'm not sure if there's a database cleaning issue here...you could check to see how many survey responses are in the database just to sanity check that there are only 2.

If you run the test in isolation, does it pass? Are the number of survey responses different if you run the test in isolation vs. running the whole file?

asheren replied:
@kopchickm It is different. When I run the whole file I get 3 surveys that are filled in with 5s and 2s like the previous tests fill out a survey. When I run just that spec in isolation, I get a blank index page with no surveys and just the headers. I messed around with truncation and transaction and even commenting out all the database cleaner code and nothing changes. It’s like the factories aren’t being called for some reason. It’s super weird and has me totally stumped and stuck. I also know that the first tr is the header. There isn’t a way to call the 2nd or last tr and I don’t think nth of a type works but I figured I need to figure out the first issue before moving on to the expectations.

asheren committed:
wip commit trying to fix failing spec

kopchickm commented on app/views/survey_responses/index.html.erb:
```diff
@@ -0,0 +1,18 @@
+    <th>Confidence Level</th>
+    <th>Difficulty of Material</th>
+    <th>Notes</th>
+  </tr>
```
I imagine the design side of things is coming later?

asheren replied:
Yup. Maybe I’ll use elm 😂

kopchickm commented on app/views/survey_responses/index.html.erb:
```diff
@@ -0,0 +1,18 @@
+
+<% @survey_responses.each do |survey_response| %>
+  <tr>
+    <td><%= survey_response.created_at.in_time_zone('Eastern Time (US & Canada)').strftime("%b %e, %Y") %></td>
```
If you find yourself doing this in multiple places, it might be nice to refactor it into a method on either SurveyResponse or into a decorator. A decorator version might look something like this:

```
def created_at
  object.created_at.in_time_zone('Eastern Time (US & Canada)').strftime("%b %e, %Y")
end
```

kopchickm commented on spec/features/survey_response_new_survey_spec.rb:
```diff
@@ -0,0 +1,74 @@
+require "rails_helper"
+
+feature "Submitting a new code confidence survey", js: true do
```
Do these tests need javascript? Just checking.

asheren replied:
No but was seeing if it made a difference with the issue I was having with the last spec. I can delete it now.

kopchickm commented on spec/features/survey_response_new_survey_spec.rb:
```diff
@@ -0,0 +1,74 @@
+
+    select "5", :from => "How confident am I"
+
+    expect(page).to have_content("5")
```
I think this test could probably be combined with the test below it to exercise a bit more code than it currently does. Just making sure that selecting an item doesn't really test your code out, since it's the built in Rails helpers that are responsible for making sure the select works.

It might be nice to combine this test with the other. Select both of the options, then submit the form and verify that the SurveyResponse was actually created, either through the UI or directly checking in the database if you need to.

kopchickm commented on spec/features/survey_response_new_survey_spec.rb:
```diff
@@ -0,0 +1,74 @@
+
+    click_button "Submit"
+
+    expect(page).to have_content("Index")
```
This test doesn't really exercise anything super useful to me. It seems like it's just testing that the redirect happens correctly, which is just a single `expect` that could be attached to a different test that submits the form.

kopchickm commented on spec/models/survey_response_spec.rb:
```diff
@@ -0,0 +1,43 @@
+RSpec.describe SurveyResponse, type: :model do
+  it "validates that perceived code confidence is answered" do
+    survey_response = build(:survey_response, perceived_code_confidence: nil)
+    survey_response.valid?
```
You don't actually need to call `survey_response.valid?` here because the next line will call it for you :) Same for the other tests in this file.

kopchickm commented on spec/support/factory_bot.rb:
```diff
@@ -0,0 +1,5 @@
+  config.include FactoryBot::Syntax::Methods
+end
+
+
```
Super nitpicky, but there are a couple blank lines here :) Rubocop would catch that for you.

kopchickm reviewed (changes requested):
This is looking good! I made a few comments and some suggestions about how to fix that test.

asheren committed:
Merge pull request #3 from asheren/delete_functionality

Adds ability for a user to delete a survey

asheren committed:
Merge pull request #2 from asheren/additional_links

Adds links

asheren committed:
wip continuing to solve test failures

asheren committed:
deleted unused code.

asheren committed:
only failing delete spec remains.
